runtime/vm/ValueTypeHelpers.hpp

```diff
@@ -144,13 +144,37 @@ class VM_ValueTypeHelpers {
 					}
 					break;
 				}
-				case 'L': { /* Nullable class type or interface type */
-					j9object_t lhsObject = objectAccessBarrier.inlineMixedObjectReadObject(currentThread, lhs, startOffset + result->offset);
-					j9object_t rhsObject = objectAccessBarrier.inlineMixedObjectReadObject(currentThread, rhs, startOffset + result->offset);
-
-					if (!VM_ValueTypeHelpers::acmp(currentThread, objectAccessBarrier, lhsObject, rhsObject)) {
+				case 'L': {
+					if (J9ROMFIELD_IS_NULL_RESTRICTED(result->field)) {
+						/* Null-free class type */
+						J9Class *fieldClass = findJ9ClassInFlattenedClassCache(clazz->flattenedClassCache, sigChar + 1, J9UTF8_LENGTH(signature) - 2);
 						rc = false;
-						goto done;
+
+						if (J9_IS_FIELD_FLATTENED(fieldClass, result->field)) {
+							rc = isSubstitutable(currentThread, objectAccessBarrier, lhs, rhs, startOffset + result->offset, fieldClass);
+						} else {
+							j9object_t lhsFieldObject = objectAccessBarrier.inlineMixedObjectReadObject(currentThread, lhs, startOffset + result->offset);
+							j9object_t rhsFieldObject = objectAccessBarrier.inlineMixedObjectReadObject(currentThread, rhs, startOffset + result->offset);
+
+							if (lhsFieldObject == rhsFieldObject) {
+								rc = true;
+							} else {
+								/* When unflattened, we get our object from the specified offset, then increment past the header to the first field. */
+								rc = isSubstitutable(currentThread, objectAccessBarrier, lhsFieldObject, rhsFieldObject, J9VMTHREAD_OBJECT_HEADER_SIZE(currentThread), fieldClass);
+							}
+						}
+						if (false == rc) {
+							goto done;
+						}
+					} else {
+						/* Nullable class type or interface type */
```

Comment:
> is the else block needed?

Yes, the else block is needed for nullable L types. The code in the else block is the same as previous `L` case.

> Perhaps L can look like what Q used to look like since "null restricted" doesn't add any semantic differences.

The code of `L` in the null restricted case is the same as what `Q` used to be. 

For the nullable case, the `L` code is doing `VM_ValueTypeHelpers::acmp()` that contains an additional `J9_IS_J9CLASS_VALUETYPE(lhsClass) && (rhsClass == lhsClass)` comparison before `isSubstitutable()`, while the previous `Q` case is directly calling`isSubstitutable()`. I guess they are different.

Reply:
Yes, a NULL check would be needed to collapse those statements.